### PR TITLE
vSphere: Zonal Terraform

### PR DESCRIPTION
**Pull request by jcpowermac**
This PR adds an additional terraform workflow
and variables to allow for the provisioning
of the RHCOS OVA (template), bootstrap and
control plane nodes into a zonal installation.

### CODE REVIEW

**Comment by jcpowermac** [comment]
#5911 is required to be merged first.

File: pkg/destroy/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -41,6 +42,15 @@ func Destroy(dir string) (err error) {
 		platform = typesazure.StackTerraformName
 	}
 
+
+	// TODO: change metadata to include string to check if we are zoning
```
[the remark this answers is not in the document]
Comment (jcpowermac) [reply pkg/destroy/bootstrap/bootstrap.go:46]: Yep, thanks, let me remove.

File: pkg/destroy/vsphere/vsphere.go [hunk 2]
```diff
@@ -127,8 +131,10 @@ func (o *ClusterUninstaller) deleteFolder() error {
 	// The installer should create at most one parent,
 	// the parent to the VirtualMachines.
 	// If there are more or less fail with error message.
-	if len(folderMoList) > 1 {
-		return errors.Errorf("Expected 1 Folder per tag but got %d", len(folderMoList))
+	if o.terraformPlatform != vspheretypes.ZoningTerraformName {
+		if len(folderMoList) > 1 {
+			return errors.Errorf("Expected 1 Folder per tag but got %d", len(folderMoList))
```
[the remark this answers is not in the document]
Comment (jcpowermac) [reply pkg/destroy/vsphere/vsphere.go:136]: right
N folder(s) per unique datacenter

**Comment by rvanderp3** [comment]
govet is failing due to a dependency on https://github.com/openshift/installer/pull/5911 .  LGTM once 5911 merges

File: pkg/asset/cluster/tfvars.go [hunk 3]
```diff
@@ -731,42 +731,70 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
+			controlPlaneConfigs[i] = c.Spec.ProviderSpec.Value.Object.(*machinev1beta1.VSphereMachineProviderSpec)
+			networkID, err = vsphereconfig.GetNetworkMoID(context.TODO(),
+				vim25Client,
+				finder,
+				controlPlaneConfigs[i].Workspace.Datacenter,
+				installConfig.Config.VSphere.Cluster,
+				controlPlaneConfigs[i].Network.Devices[0].NetworkName)
+			if err != nil {
+				return errors.Wrap(err, "failed to get vSphere network ID")
+			}
```
Comment (patrickdillon) [inline pkg/asset/cluster/tfvars.go:791]: I'm not following the purpose of this loop: is it just checking to make sure there is a valid `networkID` for each control plane node? It seems to change the logic, probably inconsequentially, to use the `networkID` from the last control plane node

File: pkg/tfvars/vsphere/vsphere.go [hunk 4]
```diff
@@ -76,7 +104,99 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
...
+
+	for k, v := range deploymentZone {
+		tempFolder := new(folder)
+
+		tempFolder.Datacenter = failureDomainZones[v.FailureDomain].Topology.Datacenter
+		tempFolder.Name = v.PlacementConstraint.Folder
+
+		if tempFolder.Name == "" {
+			tempFolder.Name = infraID
+			deploymentZone[k].PlacementConstraint.Folder = infraID
```
Comment (patrickdillon) [inline pkg/tfvars/vsphere/vsphere.go:130]: This does not seem like the right place to be setting this value. Should probably be in the machine pool defaults.

File: pkg/asset/cluster/tfvars.go (hunk 3 above)
Comment (jcpowermac) [reply pkg/asset/cluster/tfvars.go:791]: I had changed this section a couple times, I think it was an oops. Will fix, thanks!

File: pkg/tfvars/vsphere/vsphere.go [hunk 5]
```diff
@@ -2,8 +2,11 @@ package vsphere
 
 import (
 	"encoding/json"
+	"fmt"
 	"strings"
 
+	"github.com/openshift/installer/pkg/asset/installconfig"
```
Comment (rvanderp3) [inline pkg/tfvars/vsphere/vsphere.go:8]: should this be at the bottom of the imports with the other installer packages?

File: pkg/tfvars/vsphere/vsphere.go (hunk 4 above)
Comment (jcpowermac) [reply pkg/tfvars/vsphere/vsphere.go:129]: We are using `Folder` directly in terraform which is why it needs to be defaulted:
https://github.com/openshift/installer/pull/6064/files#diff-d0b75770b7d6a46f012401b853e416a64c1dc1abbfdc5867698a93ebf098cbbbR75

Its optional for a user but not for terraform, is machinepool defaults where that should be? 
More Zones could be defined than used.

Comment (patrickdillon) [reply pkg/tfvars/vsphere/vsphere.go:129]: > Its optional for a user but not for terraform, is machinepool defaults where that should be?

This is what I was thinking of... it seems to be consistent with what you are saying
https://github.com/openshift/installer/blob/master/pkg/types/vsphere/machinepool.go#L42-L65

Comment (jcpowermac) [reply pkg/tfvars/vsphere/vsphere.go:129]: > This is what I was thinking of... it seems to be consistent with what you are saying

is it?

I need to set the a platform-spec defined struct parameter to the InfraID, which the machinepool doesn't include.
And is infraID available when that runs?

**Comment by patrickdillon** [comment]
From a high-level perspective this code looks solid to me, and I trust @rvanderp3  and the rest of splat to cover any gaps in my vsphere knowledge. 

I am happy to give my `/approve` here, but I will reiterate: I think it would be really worthwhile to consider whether the multizone and normal vsphere terraform configs could be combined, to where the normal vsphere install is just a multizone install with a single availability zone. Once this goes in, I think it is unlikely that it would be refactored out, and would be a good idea to avoid the long-term maintenance of two largely overlapping configurations...

**Comment by jcpowermac** [comment]
While researching if we can combine the existing terraform and zonal found a bug that the newest commit should fix...testing

/test e2e-vsphere-zones

**Comment by jcpowermac** [comment]
upi image build failure :(

/test e2e-vsphere-zones

**Comment by jcpowermac** [comment]
ugh upi
/test e2e-vsphere-zones

**Comment by jcpowermac** [comment]
> From a high-level perspective this code looks solid to me, and I trust @rvanderp3 and the rest of splat to cover any gaps in my vsphere knowledge.
> 
> I am happy to give my `/approve` here, but I will reiterate: I think it would be really worthwhile to consider whether the multizone and normal vsphere terraform configs could be combined, to where the normal vsphere install is just a multizone install with a single availability zone. Once this goes in, I think it is unlikely that it would be refactored out, and would be a good idea to avoid the long-term maintenance of two largely overlapping configurations...

https://github.com/openshift/installer/pull/6279#issuecomment-1233277116

**Comment by jcpowermac** [comment]
Issues with our zoning CI environment

tested in VMC bootstrap completed.

**Comment by jcpowermac** [comment]
@rvanderp3 or @patrickdillon 
had to rebase for a conflict if I could get a lgtm please

**Comment by jcpowermac** [comment]
ugh ci :(

/retest-required

**Pull request merged**